core/src/main/scala/zio/jdbc/JdbcEncoder.scala

```diff
@@ -16,623 +16,271 @@
 }
 
 object JdbcEncoder extends JdbcEncoder0LowPriorityImplicits {
+
+  /**
+   * Restrictions:
+   * - Placeholder must contain only one "?"
+   */
+  def single[A](placeholder: String, valueToString: A => String): JdbcEncoder[A] = {
+    val questionMarkCount = placeholder.count(_ == '?')
+    if (questionMarkCount == 0) {
+      throw new IllegalArgumentException("Placeholder must contain one '?'")
+    }
+    if (questionMarkCount > 1) {
+      throw new IllegalArgumentException("Placeholder can contain only one '?'")
+    }
+    new JdbcEncoder[A] {
+
+      override def encode(value: A): SqlFragment =
+        SqlFragment(Chunk(SqlFragment.Segment.Syntax.apply(placeholder)))
+
+      override def unsafeSetValue(ps: PreparedStatement, index: Int, value: A): Int = {
+        ps.setString(index, valueToString(value))
+        index + 1
+      }
+
+      override def unsafeSetNull(ps: PreparedStatement, index: Int): Int = {
+        ps.setNull(index, Types.VARCHAR)
+        index + 1
+      }
+
+      override def sql(value: A): String = placeholder
+
+      override def prettyValuePrinter(value: A): String = valueToString(value)
+    }
+  }
+
   def apply[A]()(implicit encoder: JdbcEncoder[A]): JdbcEncoder[A] = encoder
 
-  implicit val intEncoder: JdbcEncoder[Int]                               = value => sql"$value"
-  implicit val longEncoder: JdbcEncoder[Long]                             = value => sql"$value"
-  implicit val doubleEncoder: JdbcEncoder[Double]                         = value => sql"$value"
-  implicit val charEncoder: JdbcEncoder[Char]                             = value => sql"$value"
-  implicit val stringEncoder: JdbcEncoder[String]                         = value => sql"$value"
-  implicit val booleanEncoder: JdbcEncoder[Boolean]                       = value => sql"$value"
-  implicit val bigIntEncoder: JdbcEncoder[java.math.BigInteger]           = value => sql"$value"
-  implicit val bigDecimalEncoder: JdbcEncoder[java.math.BigDecimal]       = value => sql"$value"
-  implicit val bigDecimalEncoderScala: JdbcEncoder[scala.math.BigDecimal] = value => sql"$value"
-  implicit val shortEncoder: JdbcEncoder[Short]                           = value => sql"$value"
-  implicit val floatEncoder: JdbcEncoder[Float]                           = value => sql"$value"
-  implicit val byteEncoder: JdbcEncoder[Byte]                             = value => sql"$value"
-  implicit val byteArrayEncoder: JdbcEncoder[Array[Byte]]                 = value => sql"$value"
-  implicit val byteChunkEncoder: JdbcEncoder[Chunk[Byte]]                 = value => sql"$value"
-  implicit val blobEncoder: JdbcEncoder[java.sql.Blob]                    = value => sql"$value"
-  implicit val uuidEncoder: JdbcEncoder[java.util.UUID]                   = value => sql"$value"
-
-  implicit def singleParamEncoder[A: SqlFragment.Setter]: JdbcEncoder[A] = value => sql"$value"
+  def apply[A](
+    onEncode: A => SqlFragment,
+    onSql: => String,
```

Comment:
Best to generalize this to take the `A`:
```suggestion
    onSql: A => String,
```

Reply:
Or better yet, better to derive the implementation and make it `final` in the trait so users don't have to specify this (because you can go from `A` to `SqlFragment`, and then from `SqlFragment` to `String`, automatically).